src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs

```diff
@@ -30,43 +30,74 @@ namespace Microsoft.CodeAnalysis.CodeRefactorings;
             foreach (var language in provider.Metadata.Languages)
             {
-                var orderable = new OrderableLanguageMetadata(
-                    provider.Metadata.Name, language, provider.Metadata.AfterTyped, provider.Metadata.BeforeTyped);
-                yield return new Lazy<CodeRefactoringProvider, OrderableLanguageMetadata>(() => provider.Value, orderable);
+                foreach (var documentKind in provider.Metadata.DocumentKinds)
+                {
+                    if (Enum.TryParse<TextDocumentKind>(documentKind, out var kind))
+                    {
+                        var documentExtensions = provider.Metadata.DocumentExtensions.Where(e => !string.IsNullOrEmpty(e));
+                        if (!documentExtensions.Any())
+                        {
+                            // Metadata without file extension applies to all files.
+                            documentExtensions = [""];
+                        }
+
+                        foreach (var documentExtension in documentExtensions)
+                        {
+                            var orderable = new OrderableLanguageDocumentMetadata(
+                                provider.Metadata.Name ?? "", language, kind, documentExtension, provider.Metadata.AfterTyped, provider.Metadata.BeforeTyped);
+                            yield return new Lazy<CodeRefactoringProvider, OrderableLanguageDocumentMetadata>(() => provider.Value, orderable);
+                        }
+                    }
+                }
             }
         }
     }
 
-    private ImmutableDictionary<string, Lazy<ImmutableArray<CodeRefactoringProvider>>> LanguageToProvidersMap
-        => _lazyLanguageToProvidersMap.Value;
+    private ImmutableDictionary<ProviderKey, Lazy<ImmutableArray<CodeRefactoringProvider>>> LanguageDocumentToProvidersMap
+        => _lazyLanguageDocumentToProvidersMap.Value;
 
     private ImmutableDictionary<CodeRefactoringProvider, CodeChangeProviderMetadata> RefactoringToMetadataMap
         => _lazyRefactoringToMetadataMap.Value;
 
     private ConcatImmutableArray<CodeRefactoringProvider> GetProviders(TextDocument document)
     {
         var allRefactorings = ImmutableArray<CodeRefactoringProvider>.Empty;
-        if (LanguageToProvidersMap.TryGetValue(document.Project.Language, out var lazyProviders))
+        var documentExtension = FileNameUtilities.GetExtension(document.FilePath) ?? "";
+
+        // Get providers for specific combination of language, doc kind and extension,
+        // e.g. (C#, AdditoinalDocument, .xaml)
+        var key = new ProviderKey(document.Project.Language, document.Kind, documentExtension);
+        if (LanguageDocumentToProvidersMap.TryGetValue(key, out var lazyProviders))
+        {
+            allRefactorings = lazyProviders.Value;
+        }
+
+        // Include providers which apply to all extensions, unless we picked it up because document had no extension
+        if (documentExtension != "")
```

Comment:
this structure is slightly confusing.  Might be simpler to write as
```
// Always get the providers that apply to any language
var key = new ProviderKey(document.Project.Language, document.Kind, string.Empty);
// ... get providers

var documentExtension = FileNameUtilities.GetExtension(document.FilePath);
if (documentExtension is not null)
{
    var key = new ProviderKey(document.Project.Language, document.Kind, documentExtension);
    .. get specific providers
}

```

Reply:
Good point